File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3007_maximum-number-that-sum-of-the-prices-is-less-than-or-equal-to-k/solutions/c/solution.c
#include <stdbool.h>

long long findMaximumNumber(long long k, int x) {
    long long accumulatedPrice = 0;
    long long num = 0;

    while (true) {
        num++;
        long long price = 0;
        for (long long i = x - 1; i < 64; i += x) {
            if ((num & (1LL << i)) != 0) {
                price++;
            }
        }

        accumulatedPrice += price;
        if (accumulatedPrice > k) {
            return num - 1;
        }
    }
}

File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3007_maximum-number-that-sum-of-the-prices-is-less-than-or-equal-to-k/solutions/python3/solution.py
class Solution:
    def findMaximumNumber(self, k: int, x: int) -> int:
        accumulated_price = 0
        num = 0
        
        while accumulated_price <= k:
            num += 1
            current_price = 0
            index = x
            
            # Calculate the price for the current number `num`
            while index - 1 < num.bit_length():
                if num & (1 << (index - 1)):
                    current_price += 1
                index += x
            
            # Update the accumulated price
            accumulated_price += current_price
            
            # If the next number makes it exceed k, return current number
            if accumulated_price > k:
                return num
        
        # Fallback return, technically unreachable
        return num
